### PR TITLE
feat(ai) - store: list entries

### PR DESCRIPTION
**Pull request by stepmikhaylov**
## Summary

- Adds `IStore.list_entries()` abstract method — lists files and folders under a prefix with `recursive`, `include_files`, `include_dirs`, and `glob_pattern` filtering; directory entries carry a trailing `/`
- Implements `list_entries` across all four backends: `FilesystemStore`, `MemoryStore`, `S3Store`, `AzureBlobStore`, each with integration tests
- Moves `s3.py` and `azure.py` into their own sub-packages (`store_providers/s3/`, `store_providers/azure/`) with per-provider `requirements.txt`; deps are auto-installed by `depends()` on first import
- Wires MinIO and Azurite into the Linux CI build so S3/Azure `list_entries` integration tests are mandatory on Ubuntu (docker not available on win/macOS, so tests skip cleanly there)

## Test plan

- [x] `test_store.py` — `FilesystemStore`, `MemoryStore`, `S3Store`, `AzureBlobStore` `list_entries` tests pass locally
- [x] CI Ubuntu: MinIO and Azurite steps start cleanly, env vars exported, S3/Azure tests run (not skipped)
- [x] CI win/macOS: S3/Azure tests skip (env vars absent), no failure
- [x] `S3Store` / `AzureBlobStore` tests fail the job (not skip) if their respective service step fails to start

## Summary by CodeRabbit

* **New Features**
  * Unified listing API across storage backends: recursive vs non-recursive, file/dir toggles, and name-pattern filtering.
  * S3 accepts configurable custom endpoints; S3 and Azure providers exported for direct use.
  * Path-traversal protections added across stores.

* **Tests**
  * Expanded and reorganized backend tests and integration scenarios covering listing behavior, patterns, and edge cases.

* **Chores**
  * Added commented .env template entries and CI steps to run local S3/Azure emulators; declared S3/Azure client dependencies.

### CODE REVIEW

File: packages/ai/tests/ai/account/test_store.py [hunk 1]
```diff
@@ -148,6 +129,364 @@ async def test_large_file(self, store):
...
+                raise e
+            client.create_bucket(Bucket=test_config['bucket'])
+
+        yield f'tmp_{uuid.uuid4().hex[:8]}'
+
+        paginator = client.get_paginator('list_objects_v2')
+        for page in paginator.paginate(Bucket=test_config['bucket'], Prefix='tmp_'):  # cleanup all tmp_-s to be safe
+            objects = [{'Key': obj['Key']} for obj in page.get('Contents', [])]
+            if objects:
+                client.delete_objects(Bucket=test_config['bucket'], Delete={'Objects': objects})
```
[the remark this answers is not in the document]
Comment (stepmikhaylov) [reply packages/ai/tests/ai/account/test_store.py]: The concern is not valid. The tests already run sequentially within a single run. Concurrent CI runs are not the case as dedicated MinIO and Azurite instances are setup for each run (see `.\.github\workflows\_build.yaml`).

**Review by kwit75 — CHANGES_REQUESTED** [review]
Solid shape on the list-entries work — pagination is correct across all four providers (S3 uses a `list_objects_v2` paginator, Azure's `list_blobs`/`walk_blobs` iterators follow continuation tokens — no 1000-key truncation), and filesystem/memory parity is exact. Requesting changes on two things the current tests hide, because they bite in our prod store config:

**1 — (major) off-by-one in S3 + Azure `list_entries` when the store has no URL prefix.** `prefix_part_len = key_prefix.count("/") - self._prefix.count("/")` is wrong when `self._prefix == ""`. With an empty store prefix, a recursive `list_entries("sub")` emits the *queried* directory itself (`sub/`) as a spurious entry. Every test passes only because the fixtures use `s3://bucket/tmp_<uuid>` (a 1-segment prefix), which happens to make the subtraction come out to 1. Under prod `STORE_URL=s3://rocketride-saas-dev-storage` (empty path), a FileStore-style key `users/<id>/files/sub` would leak `users/<id>/files/sub/`. (`s3/s3.py` ~L333/372, `azure/azure.py` ~L341/364.) Fix: derive base depth from the segment count of the *relative* query prefix rather than subtracting absolute-key slash counts — and please add a **no-store-prefix** test variant so it can't regress silently.

**2 — (major) traversal guard is unreachable with no store prefix.** The `raise StorageError("Path traversal detected")` in `_get_key` / `_get_blob_name` lives inside `if self._prefix:`. With an empty prefix, `list_entries("sub/../..")` normalizes to `..` and does NOT raise — `test_list_entries_prefix_traversal_protection` only passes thanks to the tmp_ fixture prefix. It fails *safe* on object stores (Prefix `../` matches nothing) so it's not an actual escape, but the guarantee the test asserts doesn't hold in prod. Move the check so it also runs when `self._prefix` is empty.

Minors (non-blocking): no `MaxKeys`/limit on recursive listing (fine now; cap before exposing to a user-facing endpoint); `list_entries` doesn't filter FileStore's `.dirmarker` sentinel (decide at the wiring PR); ruff B007/A001 shadowing in `_iter_keys` (`prefix`/`dir`); memory `_check_path` uses `".." in filename` substring (false-positives on `my..file.txt`).

Both majors are currently **dormant** (no caller — `file_store.py` is untouched), so functionality isn't at risk today — but the fix + a no-prefix test are tiny and this is exactly our prod config, so better to nail it here than land a passing-but-wrong baseline. 🙌

FYI unrelated to your code: the earlier red CI was just the leftover `Temp fail / exit 1` debug step at the old head (gone now, re-run is green-trending). Merge-order: this overlaps #1048's `_build.yaml` Test-step edit — land #1048 first, then rebase, so this also inherits `-s`.

**Comment by stepmikhaylov** [comment]
Addressed [r3340603140](https://github.com/rocketride-org/rocketride-server/pull/1044#discussion_r3340603140): updated the `name_pattern` docstring in `store.py` to clarify it is a **basename-only** glob, must not contain path separators (`/` or `\`), and that backends will raise `ValueError` for patterns that include them.

**Comment by stepmikhaylov** [comment]
@kwit75 CR addressed, please check.
* [#1048](https://github.com/rocketride-org/rocketride-server/pull/1048) merged
* [#1044](https://github.com/rocketride-org/rocketride-server/pull/1044) rebased onto develop with merged [#1048](https://github.com/rocketride-org/rocketride-server/pull/1048), no conflicts encountered
* Found bugs are fixed:
  * [#1995bcef](https://github.com/rocketride-org/rocketride-server/pull/1044/changes/1995bcef4526c0dd031698df2dc75008658c3805) all test for IO and `list_entries` applied to Azure/S3 stores w/o prefixes
  * [#e21f0cec](https://github.com/rocketride-org/rocketride-server/pull/1044/changes/e21f0cec7ce3332f7b8e0f4c40890bf612ed9fc2) Azure/S3 stores are fixed by test failures

**Review by asclearuc — APPROVED** [review]

**Review by kwit75 — APPROVED** [review]
Re-reviewed — both bugs confirmed fixed, clearing my request-changes. ✅

- **Empty-prefix off-by-one:** the new `_part_len(p) = p.count('/') + 1 if p else 0` normalizes prefix depth, so `prefix_part_len` is consistent whether `self._prefix` is set or empty. Traced both backends: `list_entries('sub')` on a no-prefix store no longer re-emits `sub/` itself (dir loop `range(2, 2)` is empty), and `list_entries('')` still emits top-level dirs. S3 (`store_root` empty-prefix handling) and Azure match.
- **Traversal guard:** the `name_pattern` `/` `\` `..` checks are now unconditional (out of `if self._prefix`), and `_get_key`/`_get_blob_name` gained the explicit `../`/`..` check — reachable on no-prefix stores now.
- **Tests:** the no-prefix store fixtures (`S3Store(url)` / Azure container-root, "no prefix") now run the full `test_list_entries_*` matrix incl. traversal + pattern protection — exactly the path the old `tmp_`-prefixed fixtures masked.

CI green on all platforms. Nice rework — merging.
